File: exercism/ruby/grade-school/grade_school.rb
require 'set'

class Student
  attr_reader :name, :grade

  def initialize(name, grade)
    @name = name
    @grade = grade
  end
end

class School
  def initialize
    @students = [] # array, or perhaps hash?
    @grades = Set.new
  end

  def add(name, grade) # enroll student in grade
    @students << Student.new(name, grade)
    @grades.add(grade)
  end

  def students(grade) # get list of all students in grade
    @students
      .select { |student| student.grade == grade }
      .map(&:name).sort
  end

  def students_by_grade # all students, by grade
    list = []
    @grades.each do |grade|
      list << {
        grade: grade,
        students: students(grade)
      }
    end
    list.sort_by { |student_group| student_group[:grade] }
  end

end

module BookKeeping
  VERSION = 3
end

File: exercism/javascript/grade-school/grade-school.js
function School() {
  this.students = [];
  this.grades = [];
}

School.prototype.add = function add(student, level) {
  this.students.push([student, level]);
  this.grades.push(level);
  this.grades.sort();
};

School.prototype.grade = function grade(level) {
  return this.students
    .filter(tuple => tuple[1] === level)
    .map(tuple => tuple[0])
    .sort((s1, s2) => {
      if (s1.toLowerCase() < s2.toLowerCase()) { return -1; }
      if (s1.toLowerCase() > s2.toLowerCase()) { return 1; }
      return 0;
    });
};

School.prototype.roster = function roster() {
  const theRoster = {};
  this.grades.forEach((level) => {
    theRoster[level] = this.grade(level);
  });
  return theRoster;
};

module.exports = School;
